src/libraries/Microsoft.Extensions.Configuration.FileExtensions/ref/Microsoft.Extensions.Configuration.FileExtensions.cs

```diff
@@ -8,8 +8,8 @@ namespace Microsoft.Extensions.Configuration
 {
     public static partial class FileConfigurationExtensions
     {
-        public static System.Action<Microsoft.Extensions.Configuration.FileLoadExceptionContext> GetFileLoadExceptionHandler(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder) { throw null; }
-        public static Microsoft.Extensions.FileProviders.IFileProvider GetFileProvider(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder) { throw null; }
+        public static System.Action<Microsoft.Extensions.Configuration.FileLoadExceptionContext>? GetFileLoadExceptionHandler(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder) { throw null; }
+        public static Microsoft.Extensions.FileProviders.IFileProvider? GetFileProvider(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder) { throw null; }
         public static Microsoft.Extensions.Configuration.IConfigurationBuilder SetBasePath(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, string basePath) { throw null; }
         public static Microsoft.Extensions.Configuration.IConfigurationBuilder SetFileLoadExceptionHandler(this Microsoft.Extensions.Configuration.IConfigurationBuilder builder, System.Action<Microsoft.Extensions.Configuration.FileLoadExceptionContext> handler) { throw null; }
```

Comment:
Shouldn't `System.Action<Microsoft.Extensions.Configuration.FileLoadExceptionContext> handler` be marked as nullable?

Reply:
There's no null check for it, but values of [`IConfigurationBuilder.Properties`](https://github.dev/dotnet/runtime/blob/dc0a7f93dc6a49016c7fa7ce18e09d285f066368/src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationBuilder.cs#L17) are non-nullable. 
Should we revisit that decision?

On the other hand there's a [null check for IFileProvider](https://github.dev/dotnet/runtime/blob/dc0a7f93dc6a49016c7fa7ce18e09d285f066368/src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationExtensions.cs#L30)

Reply:
@halter73 - you know a lot about Configuration. Do you see a reason that the `IConfigurationBuilder.Properties` dictionary can't have `null` values?

cc @maryamariyan @safern

Reply:
This reminds me of the discussion around `RateLimitLease.TryGetMetadata(...)`, so I see how having a `null` value in `Properties` could potentially be useful, but I'm not aware of anything that takes advantage of this. A `null` value would make it possible for `GetFileProvider` to return null even though it shouldn't today. 

https://github.com/dotnet/runtime/blob/c49c65df805304907185bbb4dab909672103ffef/src/libraries/Microsoft.Extensions.Configuration.FileExtensions/src/FileConfigurationExtensions.cs#L46-L51

We could change that code, but I don't see a compelling reason to make `Properties` values nullable at this point.

Reply:
> A null value would make it possible for GetFileProvider to return null even though it shouldn't today.

Note that this PR is changing `GetFileProvider` to return `IFileProvider?`. See line 30 of this file.

https://github.com/dotnet/runtime/pull/57434/files#diff-7e51ddcedfcc5de863d91e066688ee25d8b87a1d923fe5afa48fafd3e82ea370R30

Reply:
I think we should change `GetFileProvider` to throw when the `as` operation doesn't succeed and make the return value non-nullable. It looks like `FileConfigurationSource.FileProvider` will need to remain nullable since `EnsureDefaults(IConfigurationBuilder builder)` might not be called.

Reply:
> It looks like FileConfigurationSource.FileProvider will need to remain nullable since

Should it have `[DisallowNull]`?

Reply:
> Should it have [DisallowNull]?

My initial thought is "no", since a PhyiscalFileProvider will be created if the value is `null`. So it seems reasonable to allow setting the property to `null` in order to get this behavior. `null` is a valid value.

Reply:
I think `[DisallowNull]` is fine, but I can see some utility in being able to use `null` to reset to the default `PhysicalFileProvider`. It doesn't work that way today though, and I don't recall anyone asking for this.

I think if we allow it, we should just remove `builder.Properties[FileProviderKey]` if present rather than add a null value to the `IDictionary<string, object>`.

If we must put a null value in the dictionary for some reason, we should update the nullability of `IConfigurationBuilder.Properties`. `GetFileProvider` would need to be updated to fall back to the `PhysicalFileProvider` when the value exists in the dictionary but is null. With the PR as-is, I think it would still just return `null` despite the type signature. The compiler probably doesn't catch it because the dictionary values aren't supposed to be nullable.

Reply:
I just notice the `[DisallowNull]` was for `FileConfigurationSource.FileProvider` not `SetFileProvider`. I think leaving the property nullable is fine. Ignore the rest of my previous comment since `SetFileProvider` already disallows null values.